pkg/util/workqueue/metrics.go

```diff
@@ -151,3 +188,63 @@ func (m *defaultRetryMetrics) retry() {
 
 	m.retries.Inc()
 }
+
+type metricsFactory struct {
```

Comment:
I think it might be easiest for me to just show what I meant by not needing a factory variable exposed.

```go
var metricsFactory = struct {
	counterMetricProvider CounterMetricProvider
	gaugeMetricProvider   GaugeMetricProvider
	summaryMetricProvider SummaryMetricProvider
	setProviders          sync.Once
}{
	counterMetricProvider: noopProvider{},
	gaugeMetricProvider:   noopProvider{},
	summaryMetricProvider: noopProvider{},
}

func newQueueMetrics(name string) queueMetrics {
	var ret *defaultQueueMetrics
	if len(name) == 0 {
		return ret
	}
	return &defaultQueueMetrics{
		depth:                metricsFactory.gaugeMetricProvider.NewGuageMetric(name, "depth"),
		adds:                 metricsFactory.counterMetricProvider.NewCounterMetric(name, "adds"),
		latency:              metricsFactory.summaryMetricProvider.NewSummaryMetric(name, "latency"),
		workDuration:         metricsFactory.summaryMetricProvider.NewSummaryMetric(name, "duration"),
		addTimes:             map[t]time.Time{},
		processingStartTimes: map[t]time.Time{},
	}
}

func newRetryMetrics(name string) retryMetrics {
	var ret *defaultRetryMetrics
	if len(name) == 0 {
		return ret
	}
	return &defaultRetryMetrics{
		retries: metricsFactory.counterMetricProvider.NewCounterMetric(name, "retries"),
	}
}

// SetProviders sets the metrics providers of the metricsFactory.
func SetProviders(
	counterMetricProvider CounterMetricProvider,
	gaugeMetricProvider GaugeMetricProvider,
	summaryMetricProvider SummaryMetricProvider,
) {
	metricsFactory.setProviders.Do(func() {
		metricsFactory.counterMetricProvider = counterMetricProvider
		metricsFactory.gaugeMetricProvider = gaugeMetricProvider
		metricsFactory.summaryMetricProvider = summaryMetricProvider
	})
}
```